### PR TITLE
PR: Cache previous kernel attributes to speed up kernel launch (IPython console)

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

Cache a kernel to avoid waiting for a new kernel to launch.

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by ccordoba12** [comment]
Perhaps we should use the improvements added to the latest jupyter-client for this?

https://github.com/jupyter/jupyter_client/pull/712

I mean, it seems to me that you're trying to achieve the same thing here.

**Comment by impact27** [comment]
I don't think `MultiKernelManager` is integrated into qtconsole, is it?

**Comment by ccordoba12** [comment]
> I don't think MultiKernelManager is integrated into qtconsole, is it?

No, it isn't. But I think we should integrate it in Spyder to simplify how we handle multiple kernels for our consoles. And, as a side effect, we could benefit from the improvements I mentioned in the PR above.

**Comment by impact27** [comment]
That is a good idea! What do you want to do with this PR? It is ready and works as expected, but the slow Win test fails because `spyder.plugins.tours.widgets.FadingTipBox`gets the focus. I guess this PR trips this because of a race issue.

**Comment by impact27** [comment]
@ccordoba12 I have seen failures caused by `spyder.plugins.tours.widgets.FadingTipBox` in other PR so this is unrelated to this one

**Comment by ccordoba12** [comment]
> What do you want to do with this PR?

How substantial is the improvement with it? I mean, do you think it's worth to have it for Spyder 5, before doing the refactoring to use MultiKernelManager in 6?

> It is ready and works as expected, but the slow Win test fails because spyder.plugins.tours.widgets.FadingTipBoxgets the focus

You can skip that test on Windows if you want. No big deal about it (the most important thing is trying to make all tests run in at least one platform).

**Comment by impact27** [comment]
It is a big quality of life improvement in my opinion. Opening a new console goes from ~5-10s to <1s on m’a window machine

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 1]
```diff
@@ -1571,23 +1591,81 @@ def create_client_for_kernel(self):
...
+            # error
+            return new_kernel
+
+        # Check cached kernel has the same configuration as is being asked
+        if self._cached_kernel is not None:
+            (cached_spec,
+             cached_env,
+             cached_argv,
+             cached_dir,
+             cached_kernel) = self._cached_kernel
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:1614]: This is confusing: having an attribute called `self._cached_kernel` which contains a `cached_kernel` but it's different from that cached kernel.

So, I suggest to rename `self._cached_kernel` to `self._cached_kernel_properties` to make the difference between the two.

**Review by ccordoba12 — COMMENTED** [review]
This is a great improvement, thanks @impact27! Besides my comments and suggestions below, I have a general question for you:

What if have a dictionary of cached kernels (keyed by the tuple that defines a cached kernel), so that we can rely on them for special consoles (Sympy, Pylab) and regular ones? That way we wouldn't have to discard the previous cached kernel and create a new one for different kinds of consoles.

What do you think?

**Comment by impact27** [comment]
> What if we have a dictionary of cached kernels (keyed by the tuple that defines a cached kernel), so that we can rely on them for special consoles (Sympy, Pylab) and regular ones? That way we wouldn't have to discard the previous cached kernel and create a new one for different kinds of consoles.

I didn't want to have too many kernels cached without purpose. I guess 99% of the people will only use one kind.

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 2]
```diff
@@ -1470,12 +1483,18 @@ def create_new_client(self, give_focus=True, filename='', is_cython=False,
         self.master_clients += 1
         client_id = dict(int_id=str(self.master_clients),
                          str_id='A')
-        cf = self._new_connection_file()
+        cf, km, kc, stderr_obj, stdout_obj = self.get_new_kernel(
+            is_cython, is_pylab, is_sympy, std_dir=self._test_dir)
+
+        if cf is not None:
+            fault_obj = StdFile(cf, '.fault', self._test_dir)
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/main_widget.py:1490]: Isn't `self._test_dir` None if it is not filled?

**Comment by ccordoba12** [comment]
> I didn't want to have too many kernels cached without purpose. I guess 99% of the people will only use one kind.

Well, we could have five cached kernels or so. But this will be important when we add support to easily start kernels for any conda env detected in the system. The idea is to add a menu similar to `Special consoles` in Spyder 6 (or perhaps 5.3)

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 3]
```diff
@@ -1571,23 +1591,81 @@ def create_client_for_kernel(self):
...
+
+        # Check cached kernel has the same configuration as is being asked
+        if self._cached_kernel is not None:
+            (cached_spec,
+             cached_env,
+             cached_argv,
+             cached_dir,
+             cached_kernel) = self._cached_kernel
+            # Call interrupt_mode so the dict will be the same
+            kernel_spec.interrupt_mode
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/main_widget.py:1616]: I am not sure, but this property is created when first called. This is needed to get:
`cached_spec.__dict__ == kernel_spec.__dict__` to be True

**Comment by impact27** [comment]
The problem is that creating a key is not trivial. The created kernel depends on:
 - is_cython
 - is_pylab
 - is_sympy
 Plus on test specific values such as:
 -  std_dir
 - self._test_no_stderr
 
But also on:
 - self.CONFIGURATION
 Which might change at any point

**Comment by impact27** [comment]
Maybe the best approach is to merge this one and think about the best way to implement this, either with `MultiKernelManager` or with a dictionnary for different kernels

**Comment by ccordoba12** [comment]
Sure, no problem with that. I thought it could be simpler to have a dict of cached kernels.

I'm leaving this for 5.3.0 to test it manually for some time after being merged (5.2.2 will be release really soon).

**Comment by ccordoba12** [comment]
@impact27, please rebase on top of the latest `5.x` to remove the changes present here that you also did in PR #17205.

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27 for this! It's a great improvement!

**Pull request merged**